src/proto/connection.rs

```diff
@@ -17,6 +17,17 @@ use tokio::io::{AsyncRead, AsyncWrite};
 /// An H2 connection
 #[derive(Debug)]
 pub(crate) struct Connection<T, P, B: Buf = Bytes>
+where
+    P: Peer,
+{
+    /// Read / write frame values
+    codec: Codec<T, Prioritized<B>>,
+
+    inner: ConnectionInner<P, B>,
+}
+
+#[derive(Debug)]
+struct ConnectionInner<P, B: Buf = Bytes>
```

Comment:
I think most of the changes are pretty clear. In this one (and perhaps in `streams.rs`), this separation is larger and not immediately obvious. I think a small comment on the struct about the motivation would be helpful (such as which generics are we trying to keep separate).